Add headers on SendGrid emails sent via Sendgrid API

renatocron opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ yes ]
| Deprecations?                          | [ no ]
| BC breaks? (use the c.x branch)        | [ no ]
| Automated tests included?              | [ code is called on addMetadataToMail on SendGridApiMessageTest.php  ] 
| Issue(s) addressed                     | Fixes #10688

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:

This fix missing header `list-unsubscribe` when using transport set `Sendgrid API`.

New behavior to be discussed if should be merged: I'm removing `list-unsubscribe` if the `Precedence` is `Bulk`

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Load up [this PR](https://mautibox.com)
2. See steps in the #10688 



<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10690"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

renatocron committed:
Add headers on SendGrid emails sent via Sendgrid API

RCheesley commented:
@cla-bot check

Thanks for the PR @renatocron

renatocron commented:
pushed again, I think CS Fixer don't like comments inside array!

renatocron committed:
Add tests to improve coverage (mock headers object)

renatocron committed:
test if $headers is null before using it (test SendGridMailMetadataTest::testBaseMessage failed)

renatocron commented:
@RCheesley sorry, submitted again, now tests should pass (but coverage may decrease, or maybe there's already some other tests that set the header object properly)

renatocron committed:
Add types hint to SendGridMailMetadataTest::make_header

renatocron committed:
Use is_null instead of isset on SendGridMailMetadata::addMetadataToMail

RCheesley commented:
@escopecz could you assist with the PHPSTAN errors? https://github.com/mautic/mautic/pull/10690/commits/7dec3463e4703a560b37caa36c92822a668dd5e6

renatocron committed:
Remove is_null check from SendGridMailMetadata::addMetadataToMail

renatocron commented:
PHPSTAN failed again, reporting what I had stated in the commented above it 

 ```
// in prod, this is always defined, but not all tests are mocking headers
        if (false === is_null($headers)) {
```
        
I'm removing this check, as the unit test now passes with the mocked object.

kathrin-schmid committed:
Update language string (#10684)

The button that i updated is for saving changes made in the "Edit code" window and should thus say "Save", not "Edit".

escopecz committed:
Fixes 500 error in forms when render style is deactivated against 4.1 (#10697)

* Fixes issue #10453

* Adding functional test covering the bug

https://github.com/mautic/mautic/issues/10453

Co-authored-by: Rolando <rpayanm@gmail.com>

escopecz committed:
Fixing issues when using PHPSTAN with more processes (#10696)

escopecz committed:
Bumping to version 4.1.1

RCheesley committed:
Remove reference to mautibox (#10681)

tejasync committed:
Fix lock wait deleting unused ips (#10710)

* Deleted unused IPs in batches and some improvements.

* Added return type declaration to command methods and some improvements.

* Removed dump()

* Created separate methods for get and delete unused ips.

* Consumed repo methods in model.

* Updated command to use model methods.

* Added test case.

* Fix static analysis

escopecz commented on app/bundles/EmailBundle/Tests/Swiftmailer/SendGrid/Mail/SendGridMailMetadataTest.php:
```diff
@@ -12,17 +12,41 @@
+            ->willReturn($value);
+
+        return $an_header;
+    }
```
As the `Swift_Mime_Header` is just a value object with no dependencies then we can work with the real thing:
```suggestion
    private function make_header(string $key, string $value): \Swift_Mime_Header
    {
        $header = new \Swift_Mime_Headers_OpenDKIMHeader($key);
        $header->setValue($value);

        return $header;
    }
```

escopecz commented on app/bundles/EmailBundle/Tests/Swiftmailer/SendGrid/Mail/SendGridMailMetadataTest.php:
```diff
@@ -12,17 +12,41 @@
 
+        $headers = $this->getMockBuilder(\Swift_Mime_SimpleHeaderSet::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Better readability with
```suggestion
        $headers = $this->createMock(\Swift_Mime_SimpleHeaderSet::class);
```

escopecz reviewed (commented):
Thanks! 

I'm a bit confused about the test. It sets 2 headers but I don't see where it asserts the result headers. Am I missing something?

renatocron reviewed (commented):
I added two headers, one that would be included and another one that must be ignored. 

I think this could be tested after calling addMetadataToMail. I'm without the PHP env currently, I will be home in a few days, I can try to make this assertion

renatocron committed:
Update app/bundles/EmailBundle/Tests/Swiftmailer/SendGrid/Mail/SendGridMailMetadataTest.php

Co-authored-by: John Linhart <jan.linhart@acquia.com>

renatocron committed:
Update app/bundles/EmailBundle/Tests/Swiftmailer/SendGrid/Mail/SendGridMailMetadataTest.php

Co-authored-by: John Linhart <jan.linhart@acquia.com>

rahuld-dev committed:
Fix contact batch api for single id value (#10700)

Co-authored-by: John Linhart <admin@escope.cz>

renatocron committed:
Update SendGridMailMetadataTest.php

Remove use PHPUnit\Framework\MockObject\MockObject

renatocron commented:
@escopecz
waiting for workflow approval, I removed use PHPUnit\Framework\MockObject\MockObject;

abailey-dev committed:
fix(LeadTimelineEvent): keys do not match the func args, strips keys (#10663)

Fix error when using PHP 8:
{"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): array_merge() does not accept unknown named parameters}

Not sure if this is the most elegant way to address the issue but this bug has broken the site functionality.

tejasync committed:
User update role api (#10668)

* Allow user to update role via API

* Update user's role using API: test cases

* Update user's role using API: added more test cases

* Fix test cases and add more test case

* Added AbstractMauticTestCase::loginUser() method that allows login user in non-restricted areas

* Fix failing tests

Co-authored-by: fedys <miroslav.fedeles@gmail.com>
Co-authored-by: John Linhart <admin@escope.cz>

kuzmany committed:
Fix send unpublished SMS (#10577)

* Fix send unpublished SMS

* Fix send unpublished SMS for broadcast command

* Add CampaignSendSubscriberTest

* Add testSendSmsNotPublished test

* Minor change to tests

* Fixing types

* Automatic cs fix

Co-authored-by: John Linhart <admin@escope.cz>

escopecz commented:
@renatocron I'm still confused about the test:

> I'm a bit confused about the test. It sets 2 headers but I don't see where it asserts the result headers. Am I missing something?

renatocron committed:
Test if "to" header is really skipped

renatocron commented:
good morning, @escopecz

I've included a test check if after the mail was created header 'X-FOO' is set to 'Bar', and no header "to" exists

I set it to `nobody@email.com` but the real `to` should be kept, based on the data set on line 58 on new Mail.

I could also add a check if the "to" is set correctly, but I would need to access other method from SendGrid\Mail object, the "to" from parameter is pushed to personalizations, but that's somewhat not related to this test to be honest

escopecz reviewed (dismissed):
The code looks good 👍

kuzmany committed:
Fix group by If you use count columns for assets download (#10693)

* Fix group by If you use count columns

* Add unit tests

* Fix unit tests

* Remove define from tests

* Fix unit tests

* Fix unit tests

* Fix unit tests

Co-authored-by: John Linhart <admin@escope.cz>

eloimarquessilva committed:
Update grapesjs-custom.css (#10728)

Removing the background color f0f0f0 from gjs-field-radio class to improve the color constrast between radio's background and radio's options and returning the class parameters to the default of the original editor.

Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>

renatocron committed:
Cast $randomvalue to string

I don't know why my PHPUnit did not catch this locally, should be some extra parameter on the CI

renatocron commented:
I've fixed the PHPSTAN issue, CI can run again

RCheesley commented:
@renatocron please can you pull in the latest changes from the 4.1 branch? As this PR is on a protected branch (organisation rather than personal repo) we can't as maintainers merge in the latest changes. Thanks!

renatocron commented:
Just an information that I moved this PR to https://github.com/mautic/mautic/pull/10768 (but this other is not yet fully tagged)
